src/backend/renderer/damage.rs

```diff
@@ -0,0 +1,525 @@
+//!
+//!         render_context.draw(|_buffer| {
+//!             // Update the changed parts of the buffer
+//!
+//!             // Return the updated parts
+//!             vec![Rectangle::from_loc_and_size(Point::default(), (WIDTH, HEIGHT))]
+//!         });
+//!
+//!         last_update = now;
+//!     }
+//!
+//!     // Create a render element from the buffer
+//!     let location = Point::from((100.0, 100.0));
+//!     let render_element =
+//!         MemoryRenderBufferRenderElement::from_buffer(location, &memory_buffer, None, None);
+//!
+//!     // Render the output
+//!     damage_tracked_renderer
+//!         .render_output(
+//!             &mut renderer,
+//!             buffer_age,
+//!             &[render_element],
+//!             [0.8, 0.8, 0.9, 1.0],
+//!             &log,
+//!         )
+//!         .expect("failed to render the output");
+//! }
+//! ```
+
+use std::collections::VecDeque;
+
+use indexmap::IndexMap;
+
+use crate::{
+    backend::renderer::Frame,
+    output::Output,
+    utils::{Physical, Rectangle, Scale, Size, Transform},
+};
+
+use super::{
+    element::{Id, RenderElement},
+    utils::CommitCounter,
+};
+
+use super::{Renderer, Texture};
+
+#[derive(Debug, Clone, Copy)]
+struct ElementState {
+    last_commit: CommitCounter,
+    last_geometry: Rectangle<i32, Physical>,
+    last_z_index: usize,
+}
+
+#[derive(Debug, Default)]
+struct RendererState {
+    size: Option<Size<i32, Physical>>,
+    elements: IndexMap<Id, ElementState>,
+    old_damage: VecDeque<Vec<Rectangle<i32, Physical>>>,
+}
+
+/// Mode for the [`DamageTrackedRenderer`]
+#[derive(Debug, Clone)]
+pub enum DamageTrackedRendererMode {
+    /// Automatic mode based on a output
+    Auto(Output),
+    /// Static mode
+    Static {
+        /// Size of the static output
+        size: Size<i32, Physical>,
+        /// Scale of the static output
+        scale: Scale<f64>,
+        /// Transform of the static output
+        transform: Transform,
+    },
+}
+
+/// Output has no active mode
+#[derive(Debug, thiserror::Error)]
+#[error("Output has no active mode")]
+pub struct OutputNoMode;
+
+impl TryInto<(Size<i32, Physical>, Scale<f64>, Transform)> for DamageTrackedRendererMode {
+    type Error = OutputNoMode;
+
+    fn try_into(self) -> Result<(Size<i32, Physical>, Scale<f64>, Transform), Self::Error> {
+        match self {
+            DamageTrackedRendererMode::Auto(output) => Ok((
+                output.current_mode().ok_or(OutputNoMode)?.size,
+                output.current_scale().fractional_scale().into(),
+                output.current_transform(),
+            )),
+            DamageTrackedRendererMode::Static {
+                size,
+                scale,
+                transform,
+            } => Ok((size, scale, transform)),
+        }
+    }
+}
+
+/// Damage tracked renderer for a single output
+#[derive(Debug)]
+pub struct DamageTrackedRenderer {
+    mode: DamageTrackedRendererMode,
```

Comment:
Maybe hold on to the output rather than the mode?

Reply:
I am not sure what you mean, the mode has been introduced to be able to use the renderer explicitly without specifying an output.

Reply:
Yeah, that was an requirement for the api I brought up. E.g. if you want to render a region of an output for capture, you might want to use this api without having to create an entire virtual output. The mode can optionally hold on to an output and I expect most use-cases to be covered by that scenario, but this just makes the api for flexible for those, that need it.

Reply:
To add, I already use this to draw the cursor to a buffer used for the cursor plane.